lhotse/features/opensmile.py

```diff
@@ -0,0 +1,149 @@
+
+    name = "opensmile-wrapper"
+    config_type = OpenSmileConfig
+    smileExtractor = None
+    
+    def __init__(self, config: Optional[Any] = None):
+        super().__init__(config=config)
+        assert is_module_available(
+            "opensmile"
+        ), 'To use opensmile extractors, please "pip install opensmile" first.'
+        import opensmile
+        if isinstance(self.config.feature_set, str):
+            self.feature_set = opensmile.FeatureSet[self.config.feature_set]
+        else:
+            self.feature_set = self.config.feature_set
+        self.feature_level = opensmile.FeatureLevel(self.config.feature_level)
+        self.smileExtractor = opensmile.Smile(
+            feature_set=self.feature_set,
+            feature_level=self.feature_level,
+            sampling_rate=self.config.sampling_rate,
+            options=self.config.options,
+            loglevel=self.config.loglevel,
+            logfile=self.config.logfile,
+            channels=self.config.channels,
+            mixdown=self.config.mixdown,        
+            resample=self.config.resample,
+            num_workers=self.config.num_workers,
+            verbose=self.config.verbose
+        )  
+    @property        
+    def feature_names(self) -> List[str]:
+        return self.smileExtractor.feature_names
+        
+    def is_lld_or_lld_de(self)->bool:
+        from opensmile import FeatureLevel
+        return self.feature_level is FeatureLevel.LowLevelDescriptors or \
+               self.feature_level is FeatureLevel.LowLevelDescriptors_Deltas
+           
+    
+    @property
+    def frame_shift(self) -> Seconds:
+        import opensmile
+        if self.is_lld_or_lld_de() and \
+           self.feature_set in opensmile.FeatureSet.__members__.values():
+            # For all deafult opensmile configs frameshift is equal to 10 ms 
+            return 0.01
+        else:
+            raise NotImplementedError(
+                'frame_shift is not defined for Functionals \
+                feature level or for non default feature set. \
+                Defined featureset: {}'.format(self.config.feature_set))                
+                
+    def feature_dim(self, sampling_rate: int) -> int:
+        return len(self.feature_names)
+        
+    def feature_names(self) -> List[str]:
+        return self.smileExtractor.feature_names()
+
+    def extract(self, samples: np.ndarray, sampling_rate: int) -> np.ndarray:    
+        assert sampling_rate == self.config.sampling_rate
+        import opensmile
+        
+        feats = self.smileExtractor.process_signal(samples, sampling_rate=sampling_rate).to_numpy()
+        # Add last diff frames to the end of feats matrix to fit lhotse.utils.compute_num_frames
+        if self.is_lld_or_lld_de():
+            duration = np.shape(samples)[1]/sampling_rate
+            diff = compute_num_frames(duration, self.frame_shift, sampling_rate) - np.shape(feats)[0]
```

Comment:
Is it worth checking if that value is too large (e.g. have a tolerance of +/- 5 frames)? It could indicate a more serious error.

Reply:
Actually in the test there were examples with difference equal to 5. This is the result of difference in treatment of the last frame_len between Lhotse and OpenSmile. Now I print warning if abs(diff)>5.